Added option to specify cardiac period

rjrios915 opened the pull request:
Addresses #126 

Added option to specify cardiac period under simulation parameters

## Current situation
*Link any open issues or PRs related to this PR. Please ensure that all non-trivial PRs are first tracked and discussed in an existing GitHub issue or discussion.*


## Release Notes 
*Add a bullet point list summary of the feature and possible migration guides if this is a breaking change so this section can be added to the release notes.*


## Documentation
 <!-- todo developers: add documentation guide -->
*Please ensure that you properly document any additions.*


## Testing
*Please ensure that the PR meets the testing requirements set by GitHub Actions.*
 <!-- todo developers: check for coverage in GitHub Actions -->
*In addition, please ensure all modified and new code is covered by tests.*


## Code of Conduct & Contributing Guidelines 
- [x] I agree to follow the [Code of Conduct](https://github.com/SimVascular/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/SimVascular/.github/blob/main/CONTRIBUTING.md).

RickyAwesomeMan committed:
Added Reg CRL, Vessel, and Chambers

RickyAwesomeMan committed:
added CRL test cases

rjrios915 committed:
added a test case for CRL blood vessel block

rjrios915 committed:
Added option to specify cardiac period

menon-karthik commented:
@rjrios915 @JonathanPham Please make sure the tests are passing! 

Also, just wondering if it might be better to specify the cardiac cycle period within the relevant block that is driving the cardiac cycle (like the cardiac chamber)? For example, the currently implemented `ClosedLoopHeartPulmonary` block has this implemented within the block. It might make more sense to do this because the cycle period is not a parameter of the entire simulation really, it is only a parameter of the block that is driving the heart beat.

rjrios915 committed:
updated code

rjrios915 committed:
Fixes clang error on cardiac period #126

rjrios915 committed:
Fixes second clang error #126

rjrios915 committed:
Colon switched for semi-colon #126

rjrios915 committed:
Merge branch 'SimVascular:master' into RegChamber

rjrios915 committed:
Removed cardiac period

rjrios915 committed:
Merge branch 'RegChamber' of https://github.com/rjrios915/svZeroDSolver into RegChamber

rjrios915 committed:
merging RegChamber

menon-karthik commented:
@rjrios915 @JonathanPham Not sure why the tests are still failing but can you rebase this branch with the master so it it using the most recent version please?

rjrios915 committed:
Merge branch 'SimVascular:master' into period

rjrios915 commented:
Hello! All test cases except the data visualization should now be passing. As for where the cardiac period is defined, I think it could be useful to do so globally. The solver already utilizes a global period to calculate step size, number of cycles, and the time per cycle when outputting a single cardiac cycle even when there is no heart block, so I was thinking adding it to simulation parameters would make it easy for users to change the value if needed.

menon-karthik commented on src/solve/Solver.cpp:
```diff
@@ -9,6 +9,9 @@ Solver::Solver(const nlohmann::json& config) {
   DEBUG_MSG("Load model");
   this->model = std::shared_ptr<Model>(new Model());
   load_simulation_model(config, *this->model.get());
+  if (simparams.sim_cardiac_period > 0) {
```
Since we are now specifying cardiac cycle period in multiple places, add a condition to check that they are all consistent. See two examples below.

For time-varying parameters:
https://github.com/SimVascular/svZeroDSolver/blob/edd6156c07f12cf0a1fadf973768343dbba478c8/src/model/Model.cpp#L169-L175

For closed-loop heart block:
https://github.com/SimVascular/svZeroDSolver/blob/edd6156c07f12cf0a1fadf973768343dbba478c8/src/solve/SimulationParameters.cpp#L510-L517

menon-karthik commented on src/solve/SimulationParameters.cpp:
```diff
@@ -215,6 +215,7 @@ SimulationParameters load_simulation_params(const nlohmann::json& config) {
   sim_params.output_mean_only = sim_config.value("output_mean_only", false);
   sim_params.output_derivative = sim_config.value("output_derivative", false);
   sim_params.output_all_cycles = sim_config.value("output_all_cycles", false);
+  sim_params.sim_cardiac_period = sim_config.value("cardiac_period", 0.0);
```
Should the default value be 1.0? That is consistent with the code below:
https://github.com/SimVascular/svZeroDSolver/blob/edd6156c07f12cf0a1fadf973768343dbba478c8/src/model/Model.cpp#L205-L207

menon-karthik commented on src/solve/SimulationParameters.h:
```diff
@@ -52,6 +52,7 @@ struct SimulationParameters {
   // been read from config file yet.
   double sim_time_step_size{0.0};  ///< Simulation time step size
   double sim_abs_tol{0.0};         ///< Absolute tolerance for simulation
+  double sim_cardiac_period{0.0};  ///< Cardiac period
```
Same comment as above about the default value.

menon-karthik reviewed (dismissed):
@rjrios915 We should make sure your changes are consistent with what the code is currently doing. I have commented on specific changes and added links to where they might be inconsistent with the current code. Here's a quick summary:

Currently, `model->cardiac_cycle_period` is initialized as -1.0 here:
https://github.com/SimVascular/svZeroDSolver/blob/edd6156c07f12cf0a1fadf973768343dbba478c8/src/model/Model.h#L94

`model->cardiac_cycle_period` remains -1.0 until it is specified by a block, at which point we make sure that the newly specified value matches the value read previously from any other block (if any). I have provided examples of how that is checked in my comments below.  

If no value has been specified after all blocks have been read, a default value of 1.0 is set. This is in `Model::finalize` and I have also provided a link to that code block in my comments below.

Please make sure your current changes are consistent with that. Mainly, the default should be 1.0 and you should check that any newly specified value is consistent with previously specified values (if any).

rjrios915 committed:
Merge branch 'SimVascular:master' into master

rjrios915 committed:
Merge branch 'SimVascular:master' into period

rjrios915 committed:
(IN PROGRESS) working on code consistency

rjrios915 committed:
Updated Cardiac Period Implementation

rjrios915 commented:
Hello! We ran into the issue where the initialization of cardiac period to 1 within the model and simulation parameters separately cause an error when one would automatically be 1 but the other some valid non 1 value. To resolve this, both the model and parameters initialize the cardiac period to -1 unless specified elsewhere. If both values are -1, then solver.h sets the value to 1 (like what was happening in Model::Finalize). If both values are larger than 0, there is a check to make sure the values are consistent. If one is -1 and the other is greater than 0, then that value is set as the cardiac cycle. 

The other idea we had was to pass the parameters to the model constructor so this process could happen like originally when the model is initialized, but I was unsure of how to go about this.

Let me know your thoughts! All the tests should be passing.

rjrios915 committed:
Clang Update

rjrios915 committed:
Clang Update

rjrios915 committed:
Clang Update

rjrios915 committed:
Clang Update

ncdorn commented:
@rjrios915 ensure that documentation for cardiac period has been added

mrp089 commented:
What else needs to be done to merge this (if we still want to merge it)?

ncdorn commented:
We do still want to merge this. I will meet with @rjrios915 this week to finalize. Ricky - are all the tests passing and do we still need to address karthik's earlier comments?

mrp089 commented:
Sounds good! You might need another (empty) commit to run the tests again because GitHub automatically deletes the logs when they're too old.

rjrios915 committed:
Merge branch 'SimVascular:master' into master

rjrios915 committed:
Tests added

rjrios915 committed:
Added test cases and documentation

rjrios915 commented:
I addressed the previous issues by adding checks to make sure that there are no conflicting periods. I also added tests to check that the code is properly throwing errors.

rjrios915 committed:
Artifact Error

rjrios915 committed:
Artifact Fix?

rjrios915 committed:
Adds test

rjrios915 committed:
fixing test cases

rjrios915 committed:
Clang fix

rjrios915 committed:
Clang fix

ncdorn commented:
@mrp089 @menon-karthik The calibrator and solver tests for this PR are failing due to a GLIBCXX not found error, is this a dependency issue in the testing environment? Any idea why this might be happening?

menon-karthik commented:
> @mrp089 @menon-karthik The calibrator and solver tests for this PR are failing due to a GLIBCXX not found error, is this a dependency issue in the testing environment? Any idea why this might be happening?

Trying to figure this out right now. Having the same issue with some commits for the JOSS paper.

menon-karthik commented:
Update: I have a fix for this. But also want to fix #147 while I'm at it.

ncdorn commented:
Still failing the ubuntu-latest test. Do we need to do something to ensure the tests are running with the updated workflows?

menon-karthik commented:
> Still failing the ubuntu-latest test. Do we need to do something to ensure the tests are running with the updated workflows?

You have to rebase with the latest commits in master so they are included in your branch.

rjrios915 committed:
Merge pull request #1 from SimVascular/master

Specify GCC and clang-format versions in Github runners (#148)

rjrios915 committed:
updated testing

rjrios915 committed:
codechecks

mrp089 commented:
You can format your code automatically, e.g., in VS Code: https://code.visualstudio.com/docs/cpp/cpp-ide

rjrios915 committed:
codechecks

ncdorn commented:
all tests are passing here - @menon-karthik are we good to merge this? There was a new test case added to make sure we are catching mismatched cardiac period errors, maybe you want to look over that quickly.

ncdorn commented:
still need 1 more review on this to merge it I think

rjrios915 committed:
Merge branch 'master' into period

rjrios915 committed:
CMake Updates

rjrios915 committed:
CodeCheck

rjrios915 committed:
Testing Fixes

rjrios915 committed:
Model definition clean up

ncdorn committed:
Merge branch 'master' into period

ncdorn commented:
@rjrios915 It appears that this branch got mixed up with the branch in #168

rjrios915 committed:
Cleaned up period branch

rjrios915 committed:
More Cleanup

ncdorn commented on src/model/OpenLoopCoronaryBC.h:
typo in microvascular

rjrios915 committed:
clang

ncdorn reviewed (approved):
Looks good to me

ncdorn commented:
@mrp089 do you want to review this before we merge?

ncdorn commented:
@mrp089 do you care if we merge this before or after #176

ncdorn committed:
Merge branch 'master' into period

mrp089 reviewed (dismissed):
Minor: exclude your changes in the `.github` folder from this PR.

Apart from that, it looks good to me. It would be good to document the new option. Unfortunately, I have no idea where to put that after the documentation was moved away from this repository. I'll let @ktbolt sort that out.

ktbolt commented:
@mrp089 The important thing is that users can now find the documentation ! 

I think most SimVascular developers have updated the web documentation, very easy to do in fact.

rjrios915 committed:
Removed .github

rjrios915 committed:
Removed .github

rjrios915 committed:
Update .gitignore
